src/app/file-panel.js

```diff
@@ -108,14 +112,16 @@ function filepanel (appAPI, files) {
 
   var events = new EventManager()
   var element = template()
-  // TODO please do not add custom javascript objects, which have no
-  // relation to the DOM to DOM nodes
-  element.events = events
+
   fileExplorer.events.register('focus', function (path) {
     appAPI.switchToFile(path)
   })
 
-  return element
+  var api = { events: events }
+
+  container.appendChild(element)
+
+  return api
```

Comment:
ah ok - i think i get it.
* I normally always use the style showed in `FooEasy` below.
* If i need to get fancy - which i usually dont - but maybe I need to create tons ob instances in a short time and it's the performance bottleneck of my application, then i might optimize by utilizing javascripts prototype chain as shown in version `Foo3`.

**but...**
...if you for some reasons - maybe familiarity or the likes prefer to introduce some extra javascript keywords with special meaning on top of the HTML5 cross browser implemented and ES6 standard property `__proto__` ... let's say you would like to add: `new` and `this` and `prototype` property or maybe even `class` and `super` and the `constructor` property ... i guess the sky is the limit - to essentially do the same as what you can do with the `FooFast` style

... you might follow the examples in `FooTraditional` listed below instead.

They all work the same if you call them with `new`, but the explicit return also works if you forget the `new` operator. The ones with the explicit return works without all that magic introduced by special keywords of the language (like `this` or `new` and the knowledge of how `prototype`s in javascript work). ...just direct access via `__proto__` needs to be known if you need performance optimization.
I perceive it less heavy to reason about, because i don't have to guess what `this` is when i encounter it, e.g
* http://2ality.com/2014/05/this.html

or maybe just choose the correct behavior from the examples below :-)

```js
var instance

function FooEasy () {
  return { foo: 'bar' }
}
// normal use
console.log(instance = FooEasy()) // => 'bar'
// unusual use
console.log(instance instanceof FooEasy) // => false
console.log(instance = new FooEasy()) // => 'bar'
console.log(instance instanceof FooEasy) // => false


function FooFast () {
  return { foo: 'bar', __proto__: {/*put methods here*/} } // in case you really need the prototype chain
}
console.log(instance = FooFast()) // => 'bar'
console.log(instance instanceof FooFast) // => false
console.log(instance = new FooFast()) // => 'bar'
console.log(instance instanceof FooFast) // => false


function FooTraditional () {
  this.foo = 'bar'
}
console.log(instance = FooTraditional()) // => undefined
console.log(instance instanceof FooTraditional) // => false
console.log(instance = new FooTraditional()) // => 'bar'
console.log(instance instanceof FooTraditional) // => true

function FooTraditionalEmulated () {
  if (this instanceof Foo4)  return { foo: 'bar', __proto__: Foo4.prototype } // was called with 'new'
}
console.log(instance = FooTraditionalEmulated()) // => undefined
console.log(instance instanceof FooTraditionalEmulated) // => false
console.log(instance = new FooTraditionalEmulated()) // => 'bar'
console.log(instance instanceof FooTraditionalEmulated) // => true
```